core/unix/signal_linux.c

```diff
@@ -349,6 +349,36 @@ signalfd_thread_exit(dcontext_t *dcontext, thread_sig_info_t *info)
     TABLE_RWLOCK(sigfd_table, write, unlock);
 }
 
+void
+handle_pre_extended_syscall_sigmasks(dcontext_t *dcontext, kernel_sigset_t *mask,
+                                     size_t sizemask)
+{
+    thread_sig_info_t *info = (thread_sig_info_t *)dcontext->signal_field;
+    kernel_sigset_t safe_set;
+    ASSERT(sizemask == sizeof(kernel_sigset_t));
+    info->pre_syscall_app_sigprocmask = info->app_sigblocked;
+
+    if (mask != NULL && safe_read(mask, sizeof(safe_set), &safe_set)) {
+        signal_set_mask(dcontext, &safe_set);
+    }
+    /* Convert to system call that does not change the signal mask (poll, select,
+     * epoll_wait), as we've already emulated that. XXX i#2311: we may currently
+     * deliver incorrect signals, because the native sigprocmask the system call
+     * may get interrupted by may not be the same as the native app expects. In
+     * addition to this, the p* variants of above syscalls are not properly emu-
+     * lated w.r.t. their atomicity setting the sigprocmask and executing the
+     * syscall. */
+    priv_mcontext_t *mc = get_mcontext(dcontext);
+    MCXT_SYSNUM_REG(mc) = convert_to_non_sigmask_extended_syscall(dcontext);
```

Comment:
You're right.

> we should instead change the sigset param to NULL.

Why changing it at all in such cases, can we just leave it untouched to whatever value it was?

Reply:
But isn't the point of this PR to remove the signal mask from the syscall and move it to DR's emulation of the mask?  If we're going to change the native state (along the lines of the proposal in #2311) to be the real state it seems like we should do that more formally everywhere rather than let slip a break in our current invariant here.

Plus, if this mask does something like block a timer signal a client or DR is using, or block SIGSEGV, it can break DR or a client.

Reply:
Ok I see what you're saying. Changing the param to NULL effectively makes it a non p* version actually! Yes let's do that.

Reply:
Done (essentially re-factored the entire patch..)